### PR TITLE
tikv/util: support GiB and its families (#4541)

### PR DESCRIPTION
**Pull request by andrisak**
## What have you changed? (mandatory)

Added support to ReadableSize for GiB and its families of units while GB et al are still treated as before.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Unit test.

## Does this PR affect documentation (docs) or release note? (mandatory)

It should be updated but I'm not sure where in the docs it should go...

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

File: components/tikv_util/src/config.rs [hunk 1]
```diff
@@ -872,9 +866,16 @@ mod tests {
...
+            // units with binary prefixes
+            (" 0.5 PiB", PB / 2),
+            ("0.5 TiB", TB / 2),
+            ("0.5GiB ", GB / 2),
+            ("0.5MiB", MB / 2),
+            ("0.5KiB", KB / 2),
         ];
         for (src, exp) in decode_cases {
             let src = format!("s = {:?}", src);
+            println!("src: {:?}", src);
```
Comment (siddontang) [inline components/tikv_util/src/config.rs]: please remove the print here.

File: components/tikv_util/src/config.rs [hunk 2]
```diff
@@ -872,9 +866,16 @@ mod tests {
             ("23", 23),
             ("1", 1),
             ("1024B", KB),
+            // units with binary prefixes
+            (" 0.5 PiB", PB / 2),
```
Comment (siddontang) [inline components/tikv_util/src/config.rs]: can you add some tests with an integer number, not only float?

**Comment by siddontang** [comment]
Thanks @andrisak

**Comment by siddontang** [comment]
please format your code to fix CI

File: components/tikv_util/src/config.rs [hunk 3]
```diff
@@ -872,6 +871,17 @@ mod tests {
...
+            ("1PiB", PB),
+            ("0.5 TiB", TB / 2),
+            ("2 TiB", TB * 2),
+            ("0.5GiB ", GB / 2),
+            ("787GiB ", GB * 787),
+            ("0.5MiB", MB / 2),
+            ("3MiB", MB * 3),
+            ("0.5KiB", KB / 2),
+            ("1 KiB", KB),
```
Comment (BusyJay) [inline components/tikv_util/src/config.rs]: So “1K24B" will return "124KB" instead?

Comment (andrisak) [reply components/tikv_util/src/config.rs]: Yes.

Comment (disksing) [reply components/tikv_util/src/config.rs]: I don't think it is a good idea. Report an error would be better in this case. I think you can split str into numeric part and alphabetic part then parse them respectively.

Comment (andrisak) [reply components/tikv_util/src/config.rs]: Ok, I will update the PR.

File: components/tikv_util/src/config.rs [hunk 4]
```diff
@@ -142,35 +142,42 @@ impl FromStr for ReadableSize {
...
+            .collect::<String>();
+
+        // unit: alphabetic characters, i.e., no more digits etc expected
+        let (_, unit) = size_str.split_at(size.len());
+
+        let is_unit_invalid = unit
+            .to_string()
+            .trim()
+            .chars()
+            .find(|c| !char::is_ascii_alphabetic(c));
```
Comment (disksing) [inline components/tikv_util/src/config.rs]: The check here is redundant because `unit` will be strictly matched to the list anyway.

Comment (andrisak) [reply components/tikv_util/src/config.rs]: True, I'll fix that.

File: components/tikv_util/src/config.rs [hunk 5]
```diff
@@ -142,35 +142,32 @@ impl FromStr for ReadableSize {
...
-                if b < '0' || b > '9' {
-                    number_str = chrs.as_str();
-                    unit_char = b;
-                }
+        // size: digits and '.' as decimal separator
+        let size = size_str
+            .to_string()
+            .chars()
+            .take_while(|c| char::is_ascii_digit(c) || *c == '.')
+            .collect::<String>();
```
Comment (BusyJay) [inline components/tikv_util/src/config.rs]: Seems you can use `count` to avoid the allocation here.

File: components/tikv_util/src/config.rs [hunk 6]
```diff
@@ -142,35 +142,32 @@ impl FromStr for ReadableSize {
...
+
+        // unit: alphabetic characters
+        let (_, unit) = size_str.split_at(size.len());
```
Comment (BusyJay) [inline components/tikv_util/src/config.rs]: You can reuse the first part.

File: components/tikv_util/src/config.rs (hunk 5 above)
Comment (andrisak) [reply components/tikv_util/src/config.rs]: Good catch

**Comment by BusyJay** [comment]
Would you like to also change the serialization of ReadableSize to also use `*iB` units?

**Comment by siddontang** [comment]
PTAL @BusyJay @disksing

**Pull request merged**